MetaMask/metamask-extension
# Bump `keyring-controller` and `browser-passworder`

## **Description**

This PR bumps these package versions: 
- `@metamask/keyring-controller` to `^9.0.0` 
- `@metamask/browser-passworder` to `^4.3.0`
- `@metamask/eth-keyring-controller` to `^15.1.0`

This set of updates brings a change in the way encryption keys are derived from the user password.

To allow encryption with different iteration numbers (e.g. for different use cases), and to easily change these iterations independently in the future, an [encryptor factory](https://github.com/MetaMask/metamask-extension/pull/21878/files#diff-002d97b64bc1b560f156971f3c9dad5cfeab85f142080ee212d1bc96a2c1521a) has been added.

For performance reasons, `@metamask/eth-keyring-controller` has been _fixed_ to prefer the cached encryption key over the password (it will not derive it again), when it is initialized with `cacheEncryptionKey: true` and the encryptor supports it (which is the case with `@metamask/browser-passworder`) - for this reason, this change only impacts the login UX, instead of _every user interaction_ with the keychain.

The encryptor for Snaps has been set lower for now. 



## **Related issues**

n/a

## **Manual testing steps**

1. Unlock an existing vault should work
2. Locking and unlocking again should still work

Note from DanM: probably a good idea to test this on a lower powered machine to see the performance impact. Test the before and after impact on performance of logging in, going from the unlock screen to the home screen.

These changes might have a UX impact on slower devices, for the following actions:
- Wallet unlock
- Actions on KeyringController (or direct keyrings) that end with a call to `persistAllKeyrings`, which derives the encryption key again and re-encrypt the vault. e.g.:
  - Creating first seed phrase
  - Importing an existing seed phrase
  - ~~Add an account~~
  - ~~Remove an account~~
  - ~~Connecting a hardware wallet of any kind~~
  - ~~Adding a keyring~~
  - ~~Removing a keyring~~
  - ~~Importing account~~
  - (I might be forgetting something)
- ~~Snaps interactions: some of them call `encrypt` / `decrypt`, which also derive the key before/after all the interactions~~
- _There might be some UX oddities that we never noticed because of the fast derivation time, but that might be noticeable now ([see this as an example](https://github.com/MetaMask/metamask-extension/pull/21878#discussion_r1403470881))_

## **Screenshots/Recordings**

n/a

### **Before**


n/a

### **After**


n/a

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
